### PR TITLE
Issue #2727: Refactor System Resources OSAL Implmentation 

### PR DESCRIPTION
**Pull request by LeStarch**
| | |
|:---|:---|
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**|  |
|**_Documentation Included (y/n)_**|  |

---
## Change Description

Fixes issue #2727 by creating Os::Memory and Os::Cpu interfaces to support the non-filesystem aspects of SystemResources.

### CODE REVIEW

**Comment by LeStarch** [comment]
@thomas-bc will you review?

**Comment by LeStarch** [comment]
@bocchino will you review the use of templates here: https://github.com/nasa/fprime/blob/296aad01e2930bf20b083279fa1d556075ffaee7/Fw/Types/StringToNumber.cpp#L44-L73

I can make the argument it is a safe use of templates because: the template is not exposed to users, or outside code (it is implemented in the CPP).  The template is used as the implementation of a set of non-templated functions needed by the end user.

**Review by bocchino — APPROVED** [review]
I have reviewed the template usage. It looks good!

File: Fw/Types/StringToNumber.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
+#include <Fw/Types/StringUtils.hpp>
+#include <cstdlib>
+#include <cerrno>
+
+Fw::StringUtils::StringToNumberStatus string_to_helper_input_check(const CHAR* input, FwSizeType max_length, U8 base) {
```
Comment (thomas-bc) [inline Fw/Types/StringToNumber.cpp]: Rename `max_length` to `buffer_size`

File: Fw/Types/StringToNumber.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    // Check for null input string
+    if (input == nullptr) {
+        status = Fw::StringUtils::NULL_INPUT;
+    }
+    // Invalid base e.g. not 0, 2-36
+    else if (base == 1 || base > 36) {
+        status = Fw::StringUtils::INVALID_BASE;
+    } else {
+        // Check the string is bounded within the specified length
+        FwSizeType length = Fw::StringUtils::string_length(input, max_length);
```
Comment (thomas-bc) [inline Fw/Types/StringToNumber.cpp]: Rewrite comments for the `string_length` function and do no use / clarify `max_length`

File: Fw/Types/StringUtils.cpp [hunk 3]
```diff
@@ -1,6 +1,8 @@
 #include "StringUtils.hpp"
 #include <Fw/Types/Assert.hpp>
 #include <cstring>
+#include <cstdlib>
+#include <cerrno>
```
Comment (thomas-bc) [inline Fw/Types/StringUtils.cpp]: delete these includes

File: Os/Darwin/Cpu.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+#include <mach/mach_host.h>
+#include <mach/mach_init.h>
+#include <mach/mach_types.h>
+#include <mach/message.h>
+
+namespace Os {
+namespace Darwin {
+namespace Cpu {
+
+/**
```
Comment (thomas-bc) [inline Os/Darwin/Cpu.cpp]: Use `//!` style comments

File: Os/File.cpp [hunk 5]
```diff
@@ -233,5 +233,56 @@ File::Status File::finalizeCrc(U32 &crc) {
...
+        // EOF break out now
+        if (read == 0) {
+            size = i;
+            return Os::File::Status::OP_OK;
+        }
+        // Loop from i to i + current_chunk_size looking for `\n`
+        for (FwSignedSizeType j = i; j < (i + read); j++) {
+            // Newline seek back to after it, return the size read
+            if (buffer[j] == '\n') {
+                size = j;
```
Comment (thomas-bc) [inline Os/File.cpp]: should be `j+1`

File: Os/Linux/Cpu.cpp [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    IDLE = 4,
+    IO_WAIT = 5,
+    IRQ = 6,
+    SOFT_IRQ = 7,
+    MAX_CPU_TICK_TYPES = 8
+};
+
+
+using ProcCpuData = FwSizeType[ProcCpuMeasures::MAX_CPU_TICK_TYPES];
+constexpr FwSizeType LINE_SIZE = 256; // log10(max(U64)) * 7 rounded to nearest power of 2
```
Comment (thomas-bc) [inline Os/Linux/Cpu.cpp]: Could use some additional comments to explain what's going on and what it's going to be used for

File: Os/Stub/test/ut/StubFileTests.cpp [hunk 7]
```diff
@@ -4,10 +4,10 @@
 // ======================================================================
 #include <gtest/gtest.h>
 #include "Os/File.hpp"
-#include "Os/Models/Models.hpp"
+#include "Os/Models/Models.cpp"
```
Comment (thomas-bc) [inline Os/Stub/test/ut/StubFileTests.cpp]: remove all includes of `Models.cpp` (there's at least one other one than here)

File: cmake/platform/Linux.cmake [hunk 8]
```diff
@@ -17,6 +17,9 @@ choose_fprime_implementation(Os/Task Os/Task/Posix)
 choose_fprime_implementation(Os/Mutex Os/Mutex/Posix)
 choose_fprime_implementation(Os/Queue Os/Generic/PriorityQueue)
 
+choose_fprime_implementation(Os/Cpu Os/Stub/Cpu)
```
Comment (thomas-bc) [inline cmake/platform/Linux.cmake]: Use Linux instead of Stub

**Review by thomas-bc — CHANGES_REQUESTED** [review]

File: cmake/platform/Linux.cmake (hunk 8 above)
Comment (thomas-bc) [reply cmake/platform/Linux.cmake]: Also add CMakePresets.json

File: CMakePresets.json [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+      {
+        "name": "fprime-ninja",
+        "inherits": "fprime",
+        "displayName": "F´ Release (Ninja) Preset",
+        "description": "F´ release build using ninja and local fprime-venv",
+        "generator": "Ninja"
+      },
+      {
+        "name": "fprime-debug-ninja",
+        "inherits": "fprime-ninja",
```
Comment (thomas-bc) [inline CMakePresets.json]: Should this be `"inherits": "fprime-debug",` instead, inheriting everything from UT and adding Ninja generator?

File: CMakePresets.json [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        "displayName": "F´ Debug (Ninja) Preset",
+        "description": "F´ debug build using ninja and local fprime-venv",
+        "generator": "Ninja"
+      },
+      {
+        "name": "fprime-ut-ninja",
+        "inherits": "fprime-debug-ninja",
```
Comment (thomas-bc) [inline CMakePresets.json]: Same, `"inherits": "fprime-ut",`

**Review by thomas-bc — APPROVED** [review]

**Pull request merged**